### PR TITLE
Generate self-signed certificate and update AOS Wif.config to trust the certificate thumbprint

### PR DESCRIPTION
**Pull request by kennysaelen**
Initialize-D365TestAutomationCertificate function makes use of 2 internal functions:

- **New-D365SelfSignedCertificate**: Creates a self-signed certificate and installs it in the trusted root store

- **Add-WIFConfigAuthorityThumbprint**: Updates the wif.config file in the webroot and will add the certificate thumbprint to the authority

### CODE REVIEW

File: functions/initialize-d365testautomationcertificates.ps1 [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        [Parameter(Mandatory = $false, Position = 3)]
+        [Security.SecureString]$Password = (ConvertTo-SecureString -String "Password1" -Force -AsPlainText),
+
+        [Parameter(Mandatory = $false, Position = 4)]
+        [string]$WindowsKitFolder = "C:\Program Files (x86)\Windows Kits\10\bin\x64"
+    )
+
+    if (!$Script:IsAdminRuntime) 
+    {
+        Write-Host "The cmdlet needs administrator permission (Run As Administrator) to be able to update the configuration. Please start an elevated session and run the cmdlet again." -ForegroundColor Yellow
```
Comment (ITRasmus) [inline functions/initialize-d365testautomationcertificates.ps1]: Please use Write-PSFMessage -Level Host

File: functions/initialize-d365testautomationcertificates.ps1 [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        Write-Error "Elevated permissions needed. Please start an elevated session and run the cmdlet again." -ErrorAction Stop
```
Comment (ITRasmus) [inline functions/initialize-d365testautomationcertificates.ps1]: Please use Write-PSFMessage -Level Critical
And then Stop-PSFFunction

File: internal/functions/new-d365selfsignedcertificate.ps1 [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        $privateKeyFilePath = Join-Path $env:TEMP $PrivateKeyFileName
+        Export-PfxCertificate -cert $certificatePath -FilePath $privateKeyFilePath -Password $Password
+
+        # Import the certificate into the local machine's trusted root certificates store
+        $importedCertificate = Import-PfxCertificate -FilePath $privateKeyFilePath -CertStoreLocation Cert:\LocalMachine\Root -Password $Password    
+    }
+    catch 
+    {
+        Write-PSFMessage -Level Host -Message "Something went wrong while generating the self-signed certificate and installing it into the local machine's trusted root certificates store." -Exception $PSItem.Exception
+        Stop-PSFFunction -Message "Stopping because of errors"
```
Comment (ITRasmus) [inline internal/functions/new-d365selfsignedcertificate.ps1]: Add -StepsUpward 1 to Stop-PSFFunction

File: functions/initialize-d365testautomationcertificates.ps1 [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    try 
+    {
+        # Create the certificate and place it in the right stores
+        $X509Certificate = New-D365SelfSignedCertificate -CertificateFileName $CertificateFileName -PrivateKeyFileName $PrivateKeyFileName -Password $Password -WindowsKitFolder $WindowsKitFolder
+
```
Comment (ITRasmus) [inline functions/initialize-d365testautomationcertificates.ps1]: Use Test-PSFFunctionInterrupt to test if the function was stopped with the stop-psffunction -stepsupward 1

File: internal/functions/add-certificatethumbprintwifconfig.ps1 [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        $wifConfigFile = Join-Path ([System.Environment]::ExpandEnvironmentVariables("%ServiceDrive%")) "\AOSService\webroot\wif.config"
+        
+        [xml]$wifXml = Get-Content $wifConfigFile
+
+        $authorities = $wifXml.SelectNodes('//system.identityModel//identityConfiguration//securityTokenHandlers//securityTokenHandlerConfiguration//issuerNameRegistry//authority[@name="https://fakeacs.accesscontrol.windows.net/"]');
+        
+        if($authorities.Count -lt 1)
+        {
+            throw "Only one authority should be found with the name https://fakeacs.accesscontrol.windows.net/"
```
Comment (ITRasmus) [inline internal/functions/add-certificatethumbprintwifconfig.ps1]: Use
Write-PSFMessage 
and 
Stop-PSFFunction -Stepsupward 1

File: internal/functions/add-certificatethumbprintwifconfig.ps1 [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+               $addElem.Attributes.Append($addAtt)
+               $authority.FirstChild.AppendChild($addElem)
+               $wifXml.Save($wifConfigFile)
+            }    
+        }
+    }
+    catch 
+    {
+        Write-PSFMessage -Level Host -Message "Something went wrong while configuring the certificates and the Windows Identity Foundation configuration for the AOS" -Exception $PSItem.Exception
+        Stop-PSFFunction -Message "Stopping because of errors"
```
Comment (ITRasmus) [inline internal/functions/add-certificatethumbprintwifconfig.ps1]: Add -stepsupward 1

File: internal/functions/add-certificatethumbprintwifconfig.ps1 (hunk 5 above)
Comment (kennysaelen) [reply internal/functions/add-certificatethumbprintwifconfig.ps1]: was under the empression that throw would trigger the catch block, and there the Write-PSFMessage is present.
But added

**Comment by kennysaelen** [comment]
There are 2 extra commits now

- creating the self signed certificate has got a different provider parameter needed for the regression tool to trust the certificate
- PSF changes added as requested

**Comment by ITRasmus** [comment]
Great work Kenny.
Thanks for the added functionality.
You are correct about the Throw and Catch.
We have desided to use the Stop-PSFFunction instead of throw inside the tool there for the requested changes.
It sounds correct that there is some function we have yet to change so we use the PSFramework.
Your function is in 0.4.32 which I have just published to PSGallery.

**Pull request merged**